### PR TITLE
Changes to make Frog use Bootstrap 4

### PR DESCRIPTION
**Pull request by sleibrock**
For #185 

This PR covers the necessary changes to make Frog use Bootstrap 4. The short version of the changes are:
* upgrades files to the Bootstrap 4 bundle min distribution files (JS, CSS)
* changes to `example/_src/page-template.html` file to make adjustments for the new BS4 `navbar` class
* changes to `frog/private/bodies-page.rkt` to add pairs of tag links

The first one is obvious. The second one has changes that change the base `navbar` at the top of each page. A new function was also made in the Scribble code to better support tag links, which is what the third part is for.

Originally, when `render-template` is called to render a template page, a hash is passed with various contents, mostly strings, to be supplied to the Scribble template. This includes the fully-rendered tag links, which have HTML code included. To make changes to the `navbar`, I decided it would be best to pass in a list of pairs representing the tags, and then have the tags be rendered by the Scribble template. The list of tags looks like `'((tag-name . tag-uri) ...)`. This way only the Scribble template is responsible for rendering HTML tag links as opposed to rendering them inside of `bodies-page.rkt`.

The newly included Scribble function renders HTML links to each tag page, fully supportive of the BS4 `navbar` class. Below are some screenshots of the changes:

screenshot
screenshot2

As far as converting an existing BS3 project to BS4, I don't think there's an automatic way of doing so. I think the best way would just be copying all post and page files to a newly generated Frog project.

Also: would it be better to include jQuery into Frog as well? That's the only external file left. I could add it in as well to make Frog be completely CDN-less.

Any additional changes needed to be made I can do as well.

### CODE REVIEW

**Comment by greghendershott** [comment]
Thanks!

> As far as converting an existing BS3 project to BS4, I don't think there's an automatic way of doing so.

I think that's OK.

> Also: would it be better to include jQuery into Frog as well? That's the only external file left. I could add it in as well to make Frog be completely CDN-less.

The status quo mix is a little weird, yes.  I'm on the fence which flavor of consistent to pick -- all CDN, or no CDN.

IIRC there wasn't an "official" Bootstrap CDN when I started to Frog or I probably would have used that. I've never loved having the local copy. I appreciate there are arguments pro and con to using a CDN.

TL;DR: I'm not sure. We could discuss. Meanwhile I would accept the PR as-is (with Bootstrap local and jQuery CDN).

File: example/_src/page-template.html [hunk 1]
```diff
@@ -28,47 +28,70 @@
...
+                The pairs are in the format '((name . url) ...) }
+
+            @(define (craft-tag-links listof-pairs)
+              @(map (lambda (pair)
+                @list{
+                <a class="dropdown-item" href="@(cdr pair)">
+                  @(car pair)
+                </a>})
+              listof-pairs)
+            )
```
Comment (greghendershott) [inline example/_src/page-template.html]: This dangling closing paren jumps out at me. (But I wouldn't not merge the PR just because of this. :smile:)

Comment (sleibrock) [reply example/_src/page-template.html]: Yeah, I think I'm gonna change that section a bit.

**Comment by sleibrock** [comment]
I think it might be better to have all files included with Frog, that way sites can be served from one source alone instead of depending on CDNs (that way your site can function independently).

**Comment by greghendershott** [comment]
I don't have a strong opinion now, either way. Do you want to add a commit that includes a copy of jQuery? I'm happy to merge the PR either with or without that.

**Comment by sleibrock** [comment]
Sure, I'll push a new commit soon with the change.

**Comment by greghendershott** [comment]
Thanks!

OK, I did a very quick smoke test -- fetching the branch locally, doing a `raco frog --init`, and exploring the web site. All seems fine.

I think it's good.

I would like to squash the 6 commits down to just 1, before merging to `master`. Do you mind if I do that?

**Comment by sleibrock** [comment]
That's perfectly fine, squash away!

**Comment by greghendershott** [comment]
Oh, I just remembered the ["Project file tree"](http://docs.racket-lang.org/frog/Quick_start.html#%28part._.Project_file_tree%29) section of the docs.

This needs a few changes:
- Double-check the Bootstrap file names.
- Add the jQuery file.
- Change references to Bootstrap 3.
- Am I forgetting anything?

Is this something you would like to update, or, would you like me to?

[Scribble source](https://github.com/greghendershott/frog/blob/master/frog/frog.scrbl#L156-L231).

**Comment by greghendershott** [comment]
I need to go "offline" the rest of today. I'll close this out tomorrow. Either if you push a commit with the doc changes, or, I'll do one myself.

Thanks again for all your help!

**Comment by sleibrock** [comment]
It was my pleasure to help! I'll push the commit for the doc changes that I forgot to make (oops).

**Comment by greghendershott** [comment]
Thanks! Could you change the two occurrences of "Bootstrap 3" to "Bootstrap 4"?


1. https://github.com/sleibrock/frog/blob/a213b2df1a2c4c637681245d0632908ab1a02d1e/frog/frog.scrbl#L219

2. https://github.com/sleibrock/frog/blob/a213b2df1a2c4c637681245d0632908ab1a02d1e/frog/frog.scrbl#L223

**Comment by sleibrock** [comment]
Just changed it! I totally missed those lines (oops)

File: frog/frog.scrbl [hunk 2]
```diff
@@ -174,15 +174,14 @@ project/
     @comment{# Zero or more other @tt{.md} files, for non-post pages.}
     @comment{# May be in subdirs.}
   css/
-    bootstrap.css            @comment{# \}
-    bootstrap.min.css        @comment{# get these files}
-    bootstrap-theme.css      @comment{# from @url["http://getbootstrap.com"]}
-    bootstrap-theme.min.css  @comment{# /}
+    bootstrap.min.css        @comment{# get these files
```
Comment (greghendershott) [inline frog/frog.scrbl]: The Travis CI failure is due to the missing closing `}` on this line.

Once you add this, try running `make setup` and `make test` locally to confirm it builds and tests fine. Then when you push the commit, Travis should be all green, and we can merge!

Comment (sleibrock) [reply frog/frog.scrbl]: Just changed it! My lack of a Scribble Emacs mode may have been the problem and I didn't see this before pushing. Hopefully Travis will be fine (I see about 4 other errors reported under `make test` so hopefully that doesn't stop it)
